agents/pysmurf_controller/pysmurf_controller.py

```diff
@@ -233,33 +257,336 @@ def abort(self, session, params=None):
+            }
+            return True, "Finished taking bgmap"
+
+    @ocs_agent.param('kwargs', default=None)
+    def take_iv(self, session, params):
+        """take_iv(kwargs=None)
+
+        Takes an IV. This will add the normal resistance array and channel info
+        to the session data object along with the analyzed IV filepath.
+        See the `sodetlib IV docs page
+        <https://simons1.princeton.edu/docs/sodetlib/operations/iv.html>`_
+        for more information on what additional keyword arguments can be passed
+        in.
+
+        Args
+        ----
+        kwargs : dict
+            Additional kwargs to pass to the ``take_iv`` function.
+        """
+        if params['kwargs'] is None:
+            params['kwargs'] = {}
+
+        with self.lock.acquire_timeout(0, job='take_iv') as acquired:
+            if not acquired:
+                return False, f"Operation failed: {self.lock.job} is running."
+
+            session.set_status('starting')
+            S, cfg = self._get_smurf_control()
+            iva = iv.take_iv(S, cfg, **params['kwargs'])
+            session.data = {
+                'bands': iva.bands.tolist(),
+                'channels': iva.channels.tolist(),
+                'bgmap': iva.bgmap.tolist(),
+                'R_n': iva.R_n.tolist(),
+                'filepath': iva.filepath,
+            }
+            return True, "Finished taking IV"
+
+    @ocs_agent.param('kwargs', default=None)
+    def take_bias_steps(self, session, params):
+        """take_bias_steps(kwargs=None)
+
+        Takes bias_steps and saves the output filepath to the session data
+        object. See the `sodetlib bias step docs page
+        <https://simons1.princeton.edu/docs/sodetlib/operations/bias_steps.html>`_
+        for more information on bias steps and what kwargs can be passed in.
+
+        Args
+        ----
+        kwargs : dict
+            Additional kwargs to pass to ``take_bais_steps`` function.
+        """
+
+        if params['kwargs'] is None:
+            params['kwargs'] = {}
+
+        with self.lock.acquire_timeout(0, job='bias_steps') as acquired:
+            if not acquired:
+                return False, f"Operation failed: {self.lock.job} is running."
+
+            session.set_status('starting')
+            S, cfg = self._get_smurf_control(session=session)
+            bsa = bias_steps.take_bias_steps(
+                S, cfg, **params['kwargs']
+            )
+            session.data = {
+                'filepath': bsa.filepath
+            }
+
+            return True, "Finished taking bias steps"
+
+    @ocs_agent.param('rfrac', default=(0.3, 0.6))
+    @ocs_agent.param('kwargs', default=None)
+    def bias_dets(self, session, params):
+        """bias_dets(rfrac=(0.3, 0.6), kwargs=None)
+
+        Biases detectors to a target Rfrac value or range. This function uses
+        IV results to determine voltages for each bias-group. If rfrac is set
```

Comment:
Forgive my ignorance of sodetlib, but what IV is used?  (It looks like it doesn't acquire a new IV curve here, which is great... so does cfg know this somehow?)

Reply:
By default, whenever an IV is taken the analyzed IV output is saved and it's path is written to the detconfig object as `iv_file`. So by default bias_dets uses the most recent IV, but can take an analyzed IV object as input to use as a keyword argument. It's a simple change to allow these functions to also optionally take a path to an analyzed IV file as a kwarg, but that would be a change in the sodetlib repo, and wouldn't require modification to this task since we're already passing arbitrary kwargs through.